### PR TITLE
[BEAM-7197] ensure DoFn is torn down after exception thrown in lifecycle method

### PR DESCRIPTION
**Pull request by adude3141**
Caused by a bug in test implementation, missing calls to DoFns teardown method were hidden.

This PR
- fixes the bug in ParDoLifeCycleTest
- implements missing calls to teardown revealed be the fixed tests
- adds tests for stateful DoFn processing (required for e.g. flink which has distinct implementations)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by adude3141** [comment]
Run Dataflow ValidatesRunner

**Comment by adude3141** [comment]
Run Dataflow PortabilityApi ValidatesRunner

**Comment by adude3141** [comment]
Run Java Flink PortableValidatesRunner Batch

**Comment by adude3141** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by adude3141** [comment]
Run Spark ValidatesRunner

**Comment by adude3141** [comment]
Run Samza ValidatesRunner

**Comment by adude3141** [comment]
Run Flink ValidatesRunner

**Comment by adude3141** [comment]
Run Gearpump ValidatesRunner

**Comment by adude3141** [comment]
Run Apex ValidatesRunner

**Comment by adude3141** [comment]
Run Dataflow Portability ExecutableStage ValidatesRunner

**Comment by adude3141** [comment]
Run Flink ValidatesRunner

**Comment by adude3141** [comment]
Run Flink ValidatesRunner

**Comment by adude3141** [comment]
Run Spark ValidatesRunner

**Comment by adude3141** [comment]
Run Java Flink PortableValidatesRunner Batch

**Comment by adude3141** [comment]
Run Java Flink PortableValidatesRunner Streaming

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java [hunk 1]
```diff
@@ -36,4 +36,26 @@
...
+  public static <InputT, OutputT> DoFnInvoker<InputT, OutputT> tryInvokeSetupFor(
+      DoFn<InputT, OutputT> fn) {
+    DoFnInvoker<InputT, OutputT> doFnInvoker = invokerFor(fn);
+    try {
+      doFnInvoker.invokeSetup();
+    } catch (Exception e) {
+      try {
+        doFnInvoker.invokeTeardown();
+      } catch (Exception ignored) {
+        // ignored, we already have an exception to throw
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java]: Would it make sense to do `e.addSuppressed(ignored)`?

Comment (adude3141) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnInvokers.java]: Sure! Thanks for pointing me to that direction.

**Comment by adude3141** [comment]
Run Flink ValidatesRunner

**Comment by adude3141** [comment]
Run Spark ValidatesRunner

**Comment by adude3141** [comment]
Run Java Flink PortableValidatesRunner Batch

**Comment by adude3141** [comment]
Run Java Flink PortableValidatesRunner Streaming

**Comment by adude3141** [comment]
Thx for your help here, @kennknowles and @iemejia

**Pull request merged**